rustdoc: normalise type/field names

its-the-shrimp opened the pull request:
Updates #100961 

- `Import` -> `Use`, to better reflect the terminology of Rust & its syntax
- `TypeBinding` -> `AssocItemConstraint`, to sync up with `clean`
- `FnDecl` -> `FunctionSignature`, because that's what it is
- `Header` -> `FunctionHeader`, because `Header` is a very word that's very heavily loaded with different meanings
- `ItemEnum::AssocType`: `default` -> `type`, because those items appear in `impl` blocks as well, where they're _not_ the "default"
- `ItemEnum::AssocConst`: `default` -> `value`, see the previous point
- `ForeignType` -> `ExternType`, because "foreign" is not the right word there
- boolean fields' names made to consistently be a phrase that can be a yes/no answer, e.g. `async` -> `is_async`

The docs of `ItemEnum::AssocType::type_` & of `ItemEnum::AssocConst::value` are also updated to be up to date with the clarification of the name of the fields

its-the-shrimp committed:
rustdoc: normalise type/field names in rustdoc-json-types/jsondoclint

its-the-shrimp commented:
r? @aDotInTheVoid

camelid commented:
Hmm, some of this feels like unnecessary churn - particularly the changes to the boolean field names. For example, I don't think `is_async` is any clearer than `async`. I also don't see the point of renaming `FnDecl` since that's what it's called in rustc AFAICT. Curious what others think though.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10240237649/job/28326778202) [(plain)](https://github.com/rust-lang/rust/commit/02919d24ead234654647abe0ec44fd3cc985763a/checks/28326778202/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.92s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [rustdoc] tests/rustdoc/cross-crate-links.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/cross-crate-links" "/checkout/tests/rustdoc/cross-crate-links.rs"
--- stderr -------------------------------
21: has check failed
21: has check failed
 `XPATH PATTERN` did not match
 //@ has 'foo/index.html' '//a[@href="../all_item_types/foreigntype.FooFType.html"]' 'FooFType'
Encountered 1 errors
------------------------------------------



---- [rustdoc] tests/rustdoc/foreigntype.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/foreigntype" "/checkout/tests/rustdoc/foreigntype.rs"
--- stderr -------------------------------
4: has check failed
 File does not exist 'foreigntype/foreigntype.ExtType.html'
 File does not exist 'foreigntype/foreigntype.ExtType.html'
     //@ has foreigntype/foreigntype.ExtType.html
 File does not exist 'foreigntype/foreigntype.ExtType.html'
 File does not exist 'foreigntype/foreigntype.ExtType.html'
     //@ has - '//a[@class="fn"]' 'do_something'
15: has check failed
 `XPATH PATTERN` did not match
 //@ has foreigntype/trait.Trait.html '//a[@class="foreigntype"]' 'ExtType'
18: has check failed
 `XPATH PATTERN` did not match
 //@ has foreigntype/index.html '//a[@class="foreigntype"]' 'ExtType'
Encountered 4 errors
------------------------------------------



---- [rustdoc] tests/rustdoc/foreigntype-reexport.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/foreigntype-reexport" "/checkout/tests/rustdoc/foreigntype-reexport.rs"
--- stderr -------------------------------
14: has check failed
14: has check failed
 File does not exist 'foreigntype_reexport/sub2/foreigntype.C.html'
         //@ has foreigntype_reexport/sub2/foreigntype.C.html
32: has check failed
 File does not exist 'foreigntype_reexport/foreigntype.C2.html'
 //@ has foreigntype_reexport/foreigntype.C2.html
35: has check failed
 `XPATH PATTERN` did not match
 //@ has foreigntype_reexport/index.html '//a[@class="foreigntype"]' 'C2'
40: has check failed
 `XPATH PATTERN` did not match
 //@ has foreigntype_reexport/index.html '//a[@class="foreigntype"]' 'C'
48: has check failed
 File does not exist 'foreigntype_reexport/foreigntype.C4.html'
 //@ has foreigntype_reexport/foreigntype.C4.html
52: has check failed
 `XPATH PATTERN` did not match
 //@ has foreigntype_reexport/index.html '//a[@class="foreigntype"]' 'C4'
Encountered 6 errors
------------------------------------------



---- [rustdoc] tests/rustdoc/intra-doc/extern-type.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/extern-type" "/checkout/tests/rustdoc/intra-doc/extern-type.rs"
--- stderr -------------------------------
27: has check failed
 File does not exist 'extern_type/foreigntype.ExternType.html'
 File does not exist 'extern_type/foreigntype.ExternType.html'
 //@ has 'extern_type/foreigntype.ExternType.html'
28: hasraw check failed
 `PATTERN` did not match
 //@ hasraw 'extern_type/fn.links_to_extern_type.html' 'href="foreigntype.ExternType.html#method.f"'
30: hasraw check failed
 `PATTERN` did not match
 //@ hasraw 'extern_type/fn.links_to_extern_type.html' 'href="foreigntype.ExternType.html#method.test"'
32: hasraw check failed
 `PATTERN` did not match
 //@ hasraw 'extern_type/fn.links_to_extern_type.html' 'href="foreigntype.ExternType.html#method.g"'
Encountered 4 errors
------------------------------------------


```

</details>

workingjubilee commented:
Hm. I agree `is_async` is not clearer than `async` but I think it **is** clearer than `async_`, which just looks like a typo.

obi1kenobi commented:
I love the effort to use clearer and more consistent terminology, and document rustdoc JSON types more thoroughly! Thanks to everyone helping this along 🙏

If possible, I'd just ask that you double-check that the name changes didn't negatively affect the size of generated rustdoc JSON files, e.g. by making an enum discriminant string substantially longer than before.

`cargo-semver-checks` is used on some crates with as many as ~250k items (e.g. `aws-sdk-ec2` or `windows`), and currently the rule of thumb for rustdoc JSON is ~1KB/item on average. When the JSON file is 250MB and we need to compare two of them to each other on CI-grade hardware, IO size and parsing time are a significant factor in overall runtime. I wouldn't ask to squeeze out every last byte — I'd just hope to avoid things like "added 10 extra bytes apiece in something that'll appear 250k times in the file" because that would be a measurable slowdown. To be clear, I don't think that's happened here, but I'd love to be extra sure.

its-the-shrimp commented:
The reason for renaming `FnDecl`, and the whole PR in general, is to be more friendly to outside users, that might not be in the know about rustc conventions, a function declaration may be understood as `fn printf(fmt: &str, ...);`,even though what the `FnDecl` is there is actually just `(fmt: &str, ...)`, which is universally understood to be a function signature

its-the-shrimp commented:
regarding boolean fields, it's fine if we end up deciding to keep their names as they are, however, another reason for this PR was to close the question of naming in rustdoc-json (hence the "Fixes" annotation at the start), thus, I believe whatever renames we happen to decide on in this PR will serve as a precedent for naming conventions for rustdoc-json in the future (I believe @aDotInTheVoid wanted to document those conventions)

fbstj commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1143,7 +1143,7 @@ fn clean_fn_decl_with_args<'tcx>(
     header: Option<&hir::FnHeader>,
     args: Arguments,
-) -> FnDecl {
+) -> FunctionSignature {
```
> The reason for renaming FnDecl, and the whole PR in general, is to be more friendly to outside users, that might not be in the know about rustc conventions, a function declaration may be understood as fn printf(fmt: &str, ...);,even though what the FnDecl is there is actually just (fmt: &str, ...), which is universally understood to be a function signature

am I right in thinking that anyone using this output will know enough rust to relate the `fn` keyword to functions, and probably the `Fn*` traits.

its-the-shrimp replied:
well, `fn` is a fundamental keyword in Rust, so `fn` <-> `function` is a trivial expectation from the user, the change from `Fn` to `Function` there is purely for consistency with the existing `Function` & `FunctionPointer` structures, I don't believe we should be swayed by rustc's naming in rustdoc-json, since rustc is definitely not intended to be public API, unlike rustdoc-json

its-the-shrimp replied:
And the change from `Decl` to `Signature` is explained in my comment a bit below

bors commented:
:umbrella: The latest upstream changes (presumably #128673) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

aDotInTheVoid reviewed (changes requested):
While there's value for making all the JSON changes in one PR (we batch many small breaking changes into 1 large one), it's not a good idea for the `clean` types. If you want to make changes to them, they can be done in later, smaller PR's that just do one thing (so are easier to review).

aDotInTheVoid commented on src/rustdoc-json-types/lib.rs:
````diff
@@ -258,19 +258,19 @@ pub struct Constant {
 /// ```
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
-pub struct TypeBinding {
+pub struct AssocItemConstraint {
````
Why this change?

its-the-shrimp replied:
a "type binding" is a confusing name, and frankly outdated, since nowadays it can also constrain a constant, and it only ever operates on associated items, by constraining them

its-the-shrimp commented:
Should I revert the changes in jsondoclint as well?

aDotInTheVoid commented:
> Should I revert the changes in jsondoclint as well?

No, jsondoclint uses the types from rustdoc_json_types, so it will need to be changed.

obi1kenobi reviewed (commented):
While I am generally in favor of the renaming of what the fields are called in _Rust_, I am strongly opposed to the new longer field names _in JSON_.

I'd like to ask that we consider not introducing the `is_` prefixes into the rustdoc JSON itself, to avoid bloating the file and causing perf regressions without any human-visible improvement.

The largest crate that uses `cargo-semver-checks` produces a 475MB rustdoc JSON file, and contains ~100k functions and ~400k trait and inherent impls. Deserializing that file is a measurable bottleneck in `cargo-semver-checks` — we deserialize two such files per run, and spend about 4-5s wall-clock on it.

The proposed changes in the JSON field names in this PR will increase the file size by about 3-4MB (~1%). There's no real offsetting benefit for this regression, since the JSON is primarily meant to be machine-readable — it isn't even pretty-printed by default. There's very little advantage to having more descriptive field names repeat thousands of times.

Let's please keep the JSON field names short?

its-the-shrimp commented:
That's valid, I was also thinking about that, and I think there's a way to avoid the tug'o'war between the JSON & the Rust repr, which is having certain fields not serialised if they contain the "default" value whatever it happens to be, e.g. `false` for booleans, I think this will bring compression benefits that'll far outweigh any size increases brought by long field names

E.g.

```rust
    #[serde(default)]
    #[serde(skip_serializing_if = "is_default")] // Function to be defined
    pub is_async: bool,
```

obi1kenobi commented:
I'd love to have both default/omitted fields and short field names as much as possible!

475MB of rustdoc JSON is _a lot_, and the scary thing is that a year ago that crate "only" produced ~250MB of rustdoc JSON. I'm not sure how much of that is due to format changes vs growth in the crate, but it's not a trend we can sustain for long either way. Short field names are a cheap way to buy a few percent, and especially leaving JSON field names unchanged by only adding the `is_` prefix in Rust not JSON feels like an easy win.

its-the-shrimp commented:
@rustbot ready
I've also renamed the `can_unwind` field in `Abi` back to `unwind` to not have the name bring any assumptions about the technical details of the ABI by the word "can", and I've reverted all the changes in the HTML output of rustdoc

its-the-shrimp commented:
Looking at it now, I'm pretty sure that omitted fields could bring size improvements of much more than the 1 percent regression brought by the `is_` prefixes, there are a lot of fields carrying booleans, options, integers which employ 0 as a special value, arrays which could be omitted completely if empty, etc., then we could also omit the ABI if it's the Rust ABI, the possibilities are boundless

obi1kenobi commented:
> Looking at it now, I'm pretty sure that omitted fields could bring size improvements of much more than the 1 percent regression brought by the `is_` prefixes

I agree, and I'm in favor of that. But I don't think it's an either-or situation, and I don't think adding the `is_` prefixes like this PR is doing is making rustdoc JSON any better off. I think it's better to not have them, regardless of whether we end up omitting fields later.

its-the-shrimp commented:
@aDotInTheVoid what's your opinion on the changes to the boolean field names?

aDotInTheVoid commented:
I think that if changing the field names to be longer causes a perf regression, that's a problem with the overall serialization/de-serialization strategy, not the type design. I'm strongly opposed to changing field names or using serde defaults to try to fix perf. I want to review this PR solely based on how good the names are.

I've opened a [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/rustdoc-json.3A.20compressed.20output) to discuss potentially compressing or using alternative serialization formats for perf-sensitive users. Please don't have any further discussion around output sizes in this PR.

aDotInTheVoid commented on src/rustdoc-json-types/lib.rs:
```diff
@@ -697,11 +705,11 @@ pub enum Abi {
 pub struct Function {
     /// Information about the function signature, or declaration.
-    pub decl: FnDecl,
+    pub decl: FunctionSignature,
```
Given this, we should probably rename the field to `signature` or `sig`

aDotInTheVoid commented on src/rustdoc-json-types/lib.rs:
```diff
@@ -1036,7 +1044,7 @@ pub struct Path {
 pub struct FunctionPointer {
     /// The signature of the function.
-    pub decl: FnDecl,
+    pub decl: FunctionSignature,
```
Ditto

aDotInTheVoid reviewed (commented):
This looks really close, thanks for working on it, and sorry for the review delay.

r=me with field name renamed

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10647777623/job/29516089181) [(plain)](https://github.com/rust-lang/rust/commit/bed6554b7e3fa6c92637eed3ae7f27a212ac68c5/checks/29516089181/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.76s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/jsondocck" "--doc-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/traits/self" "--template" "/checkout/tests/rustdoc-json/traits/self.rs"
stdout: none
--- stderr -------------------------------
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_ref")].inner.function.decl.inputs[0][0]` matched to 0 values ([]) on line 12
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_ref")].inner.function.decl.inputs[0][1].borrowed_ref.type.generic` matched to 0 values ([]) on line 13
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_ref")].inner.function.decl.inputs[0][1].borrowed_ref.lifetime` matched to 0 values ([]) on line 14
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_ref")].inner.function.decl.inputs[0][1].borrowed_ref.mutable` matched to 0 values ([]) on line 15
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_exclusive_ref")].inner.function.decl.inputs[0][0]` matched to 0 values ([]) on line 18
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_exclusive_ref")].inner.function.decl.inputs[0][1].borrowed_ref.type.generic` matched to 0 values ([]) on line 19
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_exclusive_ref")].inner.function.decl.inputs[0][1].borrowed_ref.lifetime` matched to 0 values ([]) on line 20
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_exclusive_ref")].inner.function.decl.inputs[0][1].borrowed_ref.mutable` matched to 0 values ([]) on line 21
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_value")].inner.function.decl.inputs[0][0]` matched to 0 values ([]) on line 24
Failed check: Expected 3 values, but `$.index[*][?(@.name=="by_value")].inner.function.decl.inputs[0][1].generic` matched to 0 values ([]) on line 25
Failed check: Expected 3 values, but `$.index[*][?(@.name=="with_lifetime")].inner.function.decl.inputs[0][0]` matched to 0 values ([]) on line 28
Failed check: Expected 3 values, but `$.index[*][?(@.name=="with_lifetime")].inner.function.decl.inputs[0][1].borrowed_ref.type.generic` matched to 0 values ([]) on line 29
Failed check: Expected 3 values, but `$.index[*][?(@.name=="with_lifetime")].inner.function.decl.inputs[0][1].borrowed_ref.lifetime` matched to 0 values ([]) on line 30
Failed check: Expected 3 values, but `$.index[*][?(@.name=="with_lifetime")].inner.function.decl.inputs[0][1].borrowed_ref.mutable` matched to 0 values ([]) on line 31
Failed check: Expected 3 values, but `$.index[*][?(@.name=="build")].inner.function.decl.output.generic` matched to 0 values ([]) on line 34
Error: "Jsondocck failed for /checkout/tests/rustdoc-json/traits/self.rs"

Rustdoc Output:
status: exit status: 0
status: exit status: 0
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/traits/self" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/traits/self/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/traits/self" "--deny" "warnings" "/checkout/tests/rustdoc-json/traits/self.rs" "-A" "internal_features" "--output-format" "json" "-Zunstable-options"
stderr: none



```

</details>

aDotInTheVoid commented on src/llvm-project:
You managed to accidently update the llvm-project submodule somehow. This needs to be reverted before this can land.

its-the-shrimp commented:
@bors r=@aDotInTheVoid

bors commented:
@its-the-shrimp: :key: Insufficient privileges: Not in reviewers

aDotInTheVoid commented:
@bors r+


Thanks so much for working on this!

bors commented:
:pushpin: Commit f2696ab4d3095bb6ad6197e55855ebdf00f50b80 has been approved by `aDotInTheVoid`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=aDotInTheVoid f2696ab4d3095bb6ad6197e55855ebdf00f50b80 -->
<!-- homu: {"type":"Approved","sha":"f2696ab4d3095bb6ad6197e55855ebdf00f50b80","approver":"aDotInTheVoid","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
